README.md
# use-container-queries

A react hook that tracks a containers size and reports what range that width falls into within a list of breakpoints. This allows better responsive styling, where the user can style DOM elements based on their container, rather than the browser viewport.

Try out a demo [here](https://codesandbox.io/s/usecontainerqueries-s1l0k)

## Install

💡 Note: Since this module uses React Hooks, you'll need to have version >=16.8.0 of react and react-dom installed in your project

```sh
npm install use-container-queries --save-dev
# or
yarn add use-container-queries --dev
```

🚨 **Important**:
This hook uses the [Resize Observer](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) internally to measure the changes of size on your element. Browser support for the API is relatively wide and can be viewed [here](https://caniuse.com/resizeobserver). For now, this library does not ship with a polyfill (open to changing this) but you can include one in your project from [here](https://www.npmjs.com/package/resize-observer-polyfill)

## Quick Start

```jsx
import React from 'react';
import { useContainerQueries } from 'use-container-queries';

const breakpoints = {
    small: [0, 300],
    med: [301, 600],
    large: [601, 900],
    xl: [901],
};

function App() {
    const { ref, active, width } = useElementInView({ breakpoints });

    return (
        <div ref={ref}>
            The current width is: {width}
            This matches your breakpoint: {active}
        </div>
    );
}
```

### Background

> "Container queries allow an author to control styling based on the size of a containing element rather than the size of the user’s viewport."
> [Editors Draft - WICG](https://wicg.github.io/container-queries/)

We've been using Media Queries to handle responsive web design for years, but as we've started to adopt a more modular and component based design system, the shortcomings of this approach become obvious. A component may behave and look differently depending on where it is placed within your layout. Components should be responsive and independent of purely just the viewport size.

Container Queries aim to solve this by allowing you to observe the container's width to allow you to adjust your styles accordingly.

### How it works

The hook accepts an object of key/value pairs that signify the width ranges per breakpoint. These ranges are the 'min' and 'max' values in pixels for that breakpoint. For example, you want to set breakpoints to be notified when a containers width falls between 0-300, 301-600 and 601+. We'll label these as small, med, and large:

```js
const breakpoints = {
    small: [0, 300],
    med: [301, 600],
    large: [601],
};
```

If the containers current width is 430px, the hook would return `{ active: 'med', width: 430 }`. For your largest breakpoint, you can omit the 'max' value as it is implied this breakpoint is from the minimum value and above.

### API

#### `useContainerQueries(options)`

```ts
function useContainerQueries<T extends HTMLElement>(
    options: ContainerQueryProps
): ContainerQueryResult<T>;
```

#### Options: `ContainerQueryProps`

| Name             | Type                               | Default | Required | Description                                                                                                                                                                           |
| ---------------- | ---------------------------------- | ------- | -------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| breakpoints      | `[key: string]: [number, number?]` |         | true     | A key/value object of the breakpoints you wish to use. The key will label that breakpoint range. Each value should have a min and max range (Eg { small: [0, 300], med: [301, 600]} ) |
| ignoreDimensions | `boolean`                          | true    | false    | Flag to ignore updating state for container width changes when they don't trigger a breakpoint change, thus saving renders. Width will always return 0                                |

#### Returns `ContainerQueryResult<T>`

| Name   | Type             | Description                                                                                                          |
| ------ | ---------------- | -------------------------------------------------------------------------------------------------------------------- |
| ref    | `RefCallback<T>` | A callback ref to be assigned to the element you wish to observe changes on. This must be assigned                   |
| width  | `number`         | The current width of the observed element                                                                            |
| active | `string`         | The current active breakpoint. This key will match a property from the breakpoints supplied to the hook by the user. |

### Examples

Coming soon


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

-   Add ability for user to supply their own ref to the hook
-   Set `boxType` as a config option for user to choose between border and content box (ResizeObserver v2 options)

## [1.0.0] - 2021-02-19

### Added

-   Initial release of library


src/__tests__/test-utils.ts
// Store the element with the entry callback to be able to test/observe the node
// the observe method was called on.
export const observerMap = new Map<ResizeObserver, ObserverItem>();

// Mock the callback and the entry with only the properties we wish to test for brevity.
export interface MockedResizeObserverCallback {
    (entries: ResizeObserverEntry[], observer?: ResizeObserver): void;
}

interface MockedResizeObserverEntry {
    readonly target: Element;
    readonly contentRect?: DOMRectReadOnly;
    readonly borderBoxSize: ResizeObserverSize[] | ResizeObserverSize;
}

type ObserverItem = {
    callback: ResizeObserverCallback;
    element: Set<Element>;
};

// This function is used to mock the callback triggered when the intersection observer fires a change on the observed element
// We can mock the return type, to test against the isIntersecting property along with thresholds.
export function triggerObserverCallback({
    target,
    contentRect = defaultDomRect,
    borderBoxSize = defaultInlineSize,
}: MockedResizeObserverEntry): void {
    const observer = getMockedInstance(target);

    if (!observer) {
        throw new Error('No ResizeObserver instance found, Is the element still mounted?');
    }

    const item = observerMap.get(observer);
    if (item) {
        const entry: ResizeObserverEntry[] = [
            {
                target,
                contentRect,
                borderBoxSize,
                contentBoxSize: defaultInlineSize,
                devicePixelContentBoxSize: defaultInlineSize,
            },
        ];

        item.callback(entry, observer);
    }
}

export function getMockedInstance(element: Element): ResizeObserver {
    for (const [observer, item] of observerMap) {
        if (item.element.has(element)) {
            return observer;
        }
    }

    throw new Error(
        'Failed to find a ResizeObserver instance for the provided element. Is it still being observed?'
    );
}

const defaultDomRect: DOMRectReadOnly = {
    width: 0,
    height: 0,
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    x: 0,
    y: 0,
    toJSON: () => void {},
};

const defaultInlineSize: ResizeObserverSize = {
    inlineSize: 0,
    blockSize: 0,
};


src/v2.ts
import { RefCallback, useCallback, useMemo, useRef, useState } from 'react';
import { useElementResizeObserver, ObservedDimensions } from './use-element-resize-observer';

export type QueryBreakpoints2 = {
    [key: string]: number;
};

type BreakpointRange = {
    [key: string]: [number, number?];
};

export type ContainerQueryProps = {
    valueType: 'min' | 'max';
    breakpoints: QueryBreakpoints2[];
    implicitBreakpointKey?: string;
    ignoreDimensions?: boolean;
    // mode?: 'horizontal' | 'vertical'
};

type ContainerQueryState = {
    active: string;
    dimensions: ObservedDimensions;
};

export type ContainerQueryResult<T> = {
    assignRef: RefCallback<T>;
    dimensions: ObservedDimensions;
    active: string;
};

export function useContainerQueries<T extends HTMLElement>({
    breakpoints,
    valueType,
    implicitBreakpointKey,
    ignoreDimensions = false,
}: ContainerQueryProps): ContainerQueryResult<T> {
    const initialBreakpoint = Object.keys(breakpoints)[0];
    const [state, setState] = useState<ContainerQueryState>({
        active: initialBreakpoint,
        dimensions: {
            width: 0,
            height: 0,
        },
    });

    const breakpointLength = useRef(Object.keys(breakpoints).length);

    const breakpointRanges: BreakpointRange = useMemo(() => {
        return breakpoints.reduce((acc, value, i) => {
            let entry: BreakpointRange = {};
            const thisKey = Object.keys(value)[0];

            // handle implicit max
            if (valueType === 'max') {
                // first value will be the implicit min - first max range
                if (i === 0) {
                    entry[thisKey] = [0, value[thisKey]];
                } else {
                    const prevBreakpoint = breakpoints[i - 1];
                    const prevKey = Object.keys(prevBreakpoint)[0];
                    entry[thisKey] = [prevBreakpoint[prevKey] + 1, value[thisKey]];

                    // on the final value, we need to create the implicit max
                    if (i === breakpointLength.current - 1) {
                        Object.assign(entry, {
                            [implicitBreakpointKey || 'max']: [value[thisKey] + 1],
                        });
                    }
                }
            }

            // handle implicit min
            else {
                if (i !== breakpointLength.current - 1) {
                    const nextBreakpoint = breakpoints[i + 1];
                    const nextKey = Object.keys(nextBreakpoint)[0];

                    if (i === 0) {
                        entry[implicitBreakpointKey || 'min'] = [0, value[thisKey] - 1];
                    }
                    entry[thisKey] = [value[thisKey], nextBreakpoint[nextKey] - 1];
                }
                // final value, the single value will be the final min-width
                else {
                    entry[thisKey] = [value[thisKey]];
                }
            }
            return {
                ...acc,
                ...entry,
            };
        }, {} as BreakpointRange);
    }, [breakpoints, implicitBreakpointKey, valueType]);

    const matchBreakpoint = useCallback(
        (prevActive, { width, height }) => {
            let currentActive;

            for (const [key, [min, max]] of Object.entries(breakpointRanges)) {
                if (width >= min) {
                    if (max === undefined) {
                        currentActive = key;
                        break;
                    } else if (width <= max) {
                        currentActive = key;
                        break;
                    }
                }
            }

            return {
                active: currentActive ?? prevActive,
                dimensions: { width, height },
            };
        },
        [breakpointRanges]
    );

    const handleResize = useCallback(
        (dimensions: ObservedDimensions) => {
            if (ignoreDimensions) {
                const { active } = matchBreakpoint(state.active, dimensions);

                if (active !== state.active) {
                    setState((prev) => ({ ...prev, active }));
                }
            } else {
                setState((prev) => matchBreakpoint(prev.active, dimensions));
            }
        },
        [state.active, ignoreDimensions, matchBreakpoint]
    );

    const { assignRef } = useElementResizeObserver<T>({
        onResize: handleResize,
    });

    return {
        assignRef,
        dimensions: state.dimensions,
        active: state.active,
    } as const;
}


src/v3.ts
import { RefCallback, useCallback, useState } from 'react';
import { useElementResizeObserver, ObservedDimensions } from './use-element-resize-observer';

export type QueryBreakpoints = {
    [key: string]: [number, number?];
};

export type ContainerQueryProps = {
    breakpoints: QueryBreakpoints;
    ignoreDimensions?: boolean;
    // mode?: 'horizontal' | 'vertical'
};

type ContainerQueryState = {
    active: string;
    dimensions: ObservedDimensions;
};

export type ContainerQueryResult<T> = {
    assignRef: RefCallback<T>;
    dimensions: ObservedDimensions;
    active: string;
};

export function useContainerQueries<T extends HTMLElement>({
    breakpoints,
    ignoreDimensions = false,
}: ContainerQueryProps): ContainerQueryResult<T> {
    const initialBreakpoint = Object.keys(breakpoints)[0];
    const [state, setState] = useState<ContainerQueryState>({
        active: initialBreakpoint,
        dimensions: {
            width: 0,
            height: 0,
        },
    });

    const matchBreakpoint = useCallback(
        (prevActive, { width, height }) => {
            let currentActive;
            for (const [key, [min, max]] of Object.entries(breakpoints)) {
                if (width >= min) {
                    if (max === undefined) {
                        currentActive = key;
                        break;
                    } else if (width <= max) {
                        currentActive = key;
                        break;
                    }
                }
            }

            return {
                active: currentActive ?? prevActive,
                dimensions: { width, height },
            };
        },
        [breakpoints]
    );

    const handleResize = useCallback(
        (dimensions: ObservedDimensions) => {
            if (ignoreDimensions) {
                const { active } = matchBreakpoint(state.active, dimensions);

                if (active !== state.active) {
                    setState((prev) => ({ ...prev, active }));
                }
            } else {
                setState(matchBreakpoint(state.active, dimensions));
            }
        },
        [state.active, ignoreDimensions, matchBreakpoint]
    );

    const { assignRef } = useElementResizeObserver<T>({
        onResize: handleResize,
    });

    return {
        assignRef,
        dimensions: state.dimensions,
        active: state.active,
    } as const;
}


src/typings/resize-observer.d.ts
declare class ResizeObserver {
    constructor(callback: ResizeObserverCallback);
    observe: (target: Element, options?: ResizeObserverOptions) => void;
    unobserve: (target: Element) => void;
    disconnect: () => void;
}

type ResizeObserverBoxOptions = 'border-box' | 'content-box' | 'device-pixel-content-box';

interface ResizeObserverOptions {
    box?: ResizeObserverBoxOptions;
}

type ResizeObserverCallback = (entries: ResizeObserverEntry[], observer: ResizeObserver) => void;

declare class ResizeObserverEntry {
    readonly target: Element;
    readonly contentRect: DOMRectReadOnly;
    readonly borderBoxSize: ResizeObserverSize[] | ResizeObserverSize;
    readonly contentBoxSize: ResizeObserverSize[] | ResizeObserverSize;
    readonly devicePixelContentBoxSize: ResizeObserverSize[] | ResizeObserverSize;
}

declare class ResizeObserverSize {
    readonly inlineSize: number;
    readonly blockSize: number;
}

interface Window {
    ResizeObserver: typeof ResizeObserver;
    ResizeObserverEntry: typeof ResizeObserverEntry;
    ResizeObserverSize: typeof ResizeObserverSize;
}


src/use-element-resize-observer.ts
import { RefCallback, useCallback, useEffect, useRef, useState } from 'react';
import { useIsomorphicLayoutEffect } from './utils';

export type ObservedDimensions = {
    width: number | undefined;
    height: number | undefined;
};

type ElementResizeObserverOptions = {
    onResize?: (dimensions: ObservedDimensions) => void;
};

type ElementResizeObserverResult<T> = {
    assignRef: RefCallback<T>;
} & ObservedDimensions;

export function useElementResizeObserver<T extends HTMLElement>({
    onResize,
}: ElementResizeObserverOptions = {}): ElementResizeObserverResult<T> {
    const [dimensions, setDimensions] = useState<ObservedDimensions>({
        width: undefined,
        height: undefined,
    });

    const prevDimensions = useRef<ObservedDimensions>({
        width: undefined,
        height: undefined,
    });

    const observerInstanceRef = useRef<ResizeObserver | null>(null);
    const callbackRef = useRef<T | null>();

    const onResizeRef = useRef<ElementResizeObserverOptions['onResize'] | undefined>(undefined);
    onResizeRef.current = onResize;

    // track mounted state to ensure we don't attempt to set state on an unmounted component
    const hasUnmounted = useRef(false);
    useEffect(() => {
        return () => {
            hasUnmounted.current = true;
        };
    }, []);

    useIsomorphicLayoutEffect(() => {
        if (observerInstanceRef.current) {
            return;
        }

        const resizeCallback = ([entry]: readonly ResizeObserverEntry[]) => {
            let width;
            let height;

            if (entry.borderBoxSize) {
                width = Math.round(
                    Array.isArray(entry.borderBoxSize)
                        ? entry.borderBoxSize[0].inlineSize
                        : entry.borderBoxSize.inlineSize
                );

                height = Math.round(
                    Array.isArray(entry.borderBoxSize)
                        ? entry.borderBoxSize[0].blockSize
                        : entry.borderBoxSize.blockSize
                );
            } else {
                width = Math.round(entry.contentRect.width);
                height = Math.round(entry.contentRect.height);
            }

            if (
                prevDimensions.current.height !== height ||
                prevDimensions.current.width !== width
            ) {
                const newDimensions = { width, height };
                if (onResizeRef.current) {
                    onResizeRef.current(newDimensions);
                } else {
                    prevDimensions.current.height = height;
                    prevDimensions.current.width = width;

                    if (!hasUnmounted.current) {
                        setDimensions(newDimensions);
                    }
                }
            }
        };

        observerInstanceRef.current = new ResizeObserver(resizeCallback);

        if (callbackRef.current) {
            observerInstanceRef.current.observe(callbackRef.current, {
                box: 'border-box',
            });
        }

        return () => {
            observerInstanceRef.current?.disconnect();
            observerInstanceRef.current = null;
        };
    }, []);

    const assignRef = useCallback((node: T | null) => {
        if (callbackRef.current) {
            observerInstanceRef.current?.unobserve(callbackRef.current);
        }

        callbackRef.current = node;
        if (node) {
            observerInstanceRef.current?.observe(node);
        }
    }, []);

    return {
        assignRef,
        width: dimensions.width,
        height: dimensions.height,
    } as const;
}


src/index.ts
import { RefCallback, useCallback, useRef, useState } from 'react';
import { useIsomorphicLayoutEffect } from './utils';

export type QueryBreakpoints = {
    /** the values should be consisted of a min and max number, except for your largest breakpoint where it would be the min value and anything above */
    [key: string]: [number, number?];
};

export type ContainerQueryProps = {
    /** Explicit key/value map of the predefined breakpoints you wish to use. Each value should have a min and max range (Eg { small: [0, 300], med: [301, 600]} ) */
    breakpoints: QueryBreakpoints;
    /** Flag to ignore width updates if you are only interested in when a breakpoint changes. App will forego updating state for the width */
    ignoreDimensions?: boolean;
};

export type ContainerQueryResult<T> = {
    /** Callback ref to be assigned to the containing DOM node the user wishes to observe for changes. */
    ref: RefCallback<T>;
    /** The current 'active' breakpoint. This key will match one of the key/value pairs from the breakpoints supplied to the hook */
    active: string;
    /** Current width of the observed element */
    width: number;
};

type ContainerQueryState = {
    activeBreakpoint: string;
    width: number;
};

/**
 * useContainerQueries.
 *
 * A react hook utilizing the Resize Observer API that observes a containing elements width.
 * It matches that width with the users predefined breakpoint ranges and determines the containers size.
 * The main purpose of this is to allow web developers  to style DOM elements based on
 * the size of a containing element rather than the size of the browser viewport.
 *
 * @param {QueryBreakpoints} breakpoints A key/value mapping of explicit breakpoint ranges to be
 * tested against the current observed elements width to find a match.
 * @param {boolean} ignoreDimensions Flag stating the user doesn't care about the changing container width,
 * and only the breakpoint changes. The app will only update state on these changes.
 */
export function useContainerQueries<T extends HTMLElement>({
    breakpoints,
    ignoreDimensions = true,
}: ContainerQueryProps): ContainerQueryResult<T> {
    const initialBreakpoint = Object.keys(breakpoints)[0];
    const [state, setState] = useState<ContainerQueryState>({
        activeBreakpoint: initialBreakpoint,
        width: 0,
    });

    // Store refs for the resize observer instance, and observed element we are tracking.
    const observerRef = useRef<ResizeObserver | null>(null);
    const elementRef = useRef<T | null>(null);

    const matchBreakpoint = useCallback(
        (prevActive: string, width: number) => {
            let currentActive;
            for (const [key, [min, max]] of Object.entries(breakpoints)) {
                if (width >= min) {
                    if (max === undefined) {
                        currentActive = key;
                        break;
                    } else if (width <= max) {
                        currentActive = key;
                        break;
                    }
                }
            }

            return {
                activeBreakpoint: currentActive || prevActive,
                width,
            };
        },
        [breakpoints]
    );

    /**
     * Callback triggered on from the resize observer on change.
     */
    const handleResize = useCallback(
        ([entry]: readonly ResizeObserverEntry[]) => {
            let width: number;
            if (entry.borderBoxSize) {
                // Checking for chrome as using a non-standard array (https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver)
                width = Math.round(
                    Array.isArray(entry.borderBoxSize)
                        ? entry.borderBoxSize[0].inlineSize
                        : entry.borderBoxSize.inlineSize
                );
            } else {
                width = Math.round(entry.contentRect.width);
            }

            if (ignoreDimensions) {
                const { activeBreakpoint } = matchBreakpoint(state.activeBreakpoint, width);

                if (activeBreakpoint !== state.activeBreakpoint) {
                    setState((prev) => ({ ...prev, activeBreakpoint }));
                }
            } else {
                setState(matchBreakpoint(state.activeBreakpoint, width));
            }
        },
        [state.activeBreakpoint, ignoreDimensions, matchBreakpoint]
    );

    useIsomorphicLayoutEffect(() => {
        if (!observerRef.current) {
            observerRef.current = new ResizeObserver(handleResize);
        }

        if (elementRef.current) {
            observerRef.current.observe(elementRef.current, {
                box: 'border-box',
            });
        }

        return () => {
            observerRef.current?.disconnect();
            observerRef.current = null;
        };
    }, [handleResize]);

    // Ref callback passed to user
    // todo: allow user to provide their own ref.
    const assignRef = useCallback((node: T | null) => {
        if (elementRef.current) {
            observerRef.current?.unobserve(elementRef.current);
        }

        elementRef.current = node;
        if (node) {
            observerRef.current?.observe(node);
        }
    }, []);

    return {
        ref: assignRef,
        width: state.width,
        active: state.activeBreakpoint,
    } as const;
}
